### PR TITLE
Psionics Rework: Kyoushiki Murasaki

### LINKED ISSUE #16598: [Bug]: TechnoMancer Core bugs out after reconnecting
### Checks

- [X] I have searched the bug with a few keywords, and I confirm this bug was not yet reported.
- [X] The round I am reporting the bug from, or I am going to talk about, has already ended.

### Location

Horizon

### Describe the issue

After having to reconnect as a technomancer. My "Spell" Core Ceases to function. Even after taking it off and putting it back on. along with taking it off waiting a few seconds and putting it back on. Nothing seems to pop back up.

### How to reproduce

1. Roll Technomancer
2. Reconnect
3. Core stops working
4. Hopefully a benevolent fix.

### Round ID

coz-a9tv

### Staff ckey

_No response_

### PR DESCRIPTION
**Pull request by NonQueueingMatt**
The old psionics system has been gutted and replaced with abilities a-la Technomancer. A truckload of new abilities have been added that cover all kinds of use cases. Antagonists may now use a pointshop (opened by shift-clicking the psi hub icon) to pick and choose what psionics they want to fit their playstyle.

Additionally, adds Nlom Awareness: psionic users are prompted every 30 minutes for their emotional state. The winning emotion is then broadcasted to them.

Fixes #16598

To-do:

- [x] Balance psi costs
- [x] Psionic damage modifiers for non-psionics
- [x] Character creation-level psionic selection for Skrell
- [x] Apex abilities
- [x] Limitless abilities

### CODE REVIEW

**Comment by Butterrobber202** [comment]
We are so back

**Comment by NonQueueingMatt** [comment]
!review

**Comment by NonQueueingMatt** [comment]
!review

File: code/modules/psionics/abilities/command.dm [hunk 1]
```diff
@@ -0,0 +1,78 @@
+/singleton/psionic_power/command
+	name = "Command"
+	desc = "Send a psionic command to a target. It must be one word. The suggestion ends when they finish the task, when a minute passes or \
+			when they take notable damage."
+	icon_state = "wiz_blink"
+	spell_path = /obj/item/spell/command
+	ability_flags = PSI_FLAG_EVENT
+
+/obj/item/spell/command
+	name = "commune"
```
Comment (RustingWithYou) [inline code/modules/psionics/abilities/command.dm:10]:
```suggestion
	name = "command"
```

File: code/datums/mind.dm [hunk 2]
```diff
@@ -441,6 +436,16 @@
 		for(var/datum/objective/objective in objectives)
 			to_chat(current, "<B>Objective #[obj_count]</B>: [objective.explanation_text]")
 			obj_count++
+
+	else if(href_list["set_psi_rank"])
+		current.set_psi_rank(text2num(href_list["set_psi_rank"]))
+		return
+	else if(href_list["set_psi_rank_limitless"])
+		var/sure = input(usr, "Limitless is INTENTIONALLY STUPIDLY OVERPOWERED! YOU SHOULD NOT BE USING THIS WITHOUT KNOWING EXACTLY WHAT YOU'RE DOING!", "Don't Get A Staff Complaint") as anything in list("I know what I'm doing!", "Nevermind...")
```
Comment (Arrow768) [inline code/datums/mind.dm:444]:
```suggestion
		var/sure = input(usr, "Limitless is INTENTIONALLY STUPIDLY OVERPOWERED! YOU SHOULD NOT BE USING THIS WITHOUT KNOWING EXACTLY WHAT YOU'RE DOING!", "Don't Get A Staff Complaint") as anything in list("I know what I'm doing!", "I fear no man. But that thing, it scares me!")
```

**Comment by NonQueueingMatt** [comment]
Fixes #16598

**Review by Alberyk — APPROVED** [review]

File: code/modules/mob/living/carbon/human/human_helpers.dm [hunk 3]
```diff
@@ -197,13 +197,10 @@
...
 	to_chat(src, SPAN_DANGER("An indescribable, brain-tearing sound hisses from [source], and you collapse in a seizure!"))
 	seizure()
-	var/new_latencies = rand(2,4)
-	var/list/faculties = list(PSI_COERCION, PSI_REDACTION, PSI_ENERGISTICS, PSI_PSYCHOKINESIS)
-	for(var/i = 1 to new_latencies)
-		custom_pain(SPAN_DANGER("<font size = 3>[pick(psi_operancy_messages)]</font>"), 25)
-		set_psi_rank(pick_n_take(faculties), allow_latency ? PSI_RANK_LATENT : PSI_RANK_OPERANT) // if set to latent, it spikes anywhere from OPERANT to PARAMOUNT
-		sleep(30)
-	addtimer(CALLBACK(psi, TYPE_PROC_REF(/datum/psi_complexus, check_latency_trigger), 100, source, TRUE), 4.5 SECONDS)
+	custom_pain(SPAN_DANGER("<font size = 3>[pick(psi_operancy_messages)]</font>"), 25)
```
Comment (FluffyGhoster) [inline code/modules/mob/living/carbon/human/human_helpers.dm:200]:
```suggestion
	custom_pain(SPAN_DANGER(FONT_LARGE("[pick(psi_operancy_messages)]")), 25)
```

File: code/modules/organs/organ_external.dm [hunk 4]
```diff
@@ -284,8 +284,26 @@
 	var/laser = (damage_flags & DAMAGE_FLAG_LASER)
 	var/sharp = (damage_flags & DAMAGE_FLAG_SHARP)
 	var/edge = (damage_flags & DAMAGE_FLAG_EDGE)
+	var/psionic = (damage_flags & DAMAGE_FLAG_PSIONIC)
```
Comment (FluffyGhoster) [inline code/modules/organs/organ_external.dm:287]:
```suggestion
	var/psionic = HAS_FLAG(damage_flags, DAMAGE_FLAG_PSIONIC)
```

File: code/modules/psionics/abilities/assay.dm [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+	if(!ishuman(hit_atom))
+		return
+
+	if(!isliving(user))
+		return
+
+	var/mob/living/carbon/human/target = hit_atom
+
+	if(target.stat == DEAD || target.status_flags & FAKEDEATH)
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/assay.dm:52]:
```suggestion
	if(target.stat == DEAD || HAS_FLAG(target.status_flags, FAKEDEATH))
```

File: code/modules/psionics/abilities/command.dm [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		return
+
+	log_say("[key_name(user)] issued a [multi_word ? "multi-word" : "single-word"] command to [key_name(target)]: [text]",ckey=key_name(src))
+
+	to_chat(user, SPAN_CULT("You psionically command to [target]: [text]"))
+
+	for (var/mob/M in player_list)
+		if (istype(M, /mob/abstract/new_player))
+			continue
+		else if(M.stat == DEAD && M.client.prefs.toggles & CHAT_GHOSTEARS)
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/command.dm:67]:
```suggestion
		else if(M.stat == DEAD && HAS_FLAG(M.client.prefs.toggles, CHAT_GHOSTEARS))
```

File: code/modules/psionics/abilities/emotional_suggestion.dm [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+		return
+
+	log_say("[key_name(user)] suggested an emotion to [key_name(target)]: [text]",ckey=key_name(src))
+
+	to_chat(user, SPAN_CULT("You psionically suggest an emotion to [target]: [text]"))
+
+	for (var/mob/M in player_list)
+		if (istype(M, /mob/abstract/new_player))
+			continue
+		else if(M.stat == DEAD && M.client.prefs.toggles & CHAT_GHOSTEARS)
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/emotional_suggestion.dm:61]:
```suggestion
		else if(M.stat == DEAD && HAS_FLAG(M.client.prefs.toggles, CHAT_GHOSTEARS))
```

File: code/modules/psionics/abilities/grip.dm [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+/obj/item/spell/grip/on_use_cast(mob/user, bypass_psi_check)
+	if(!victim)
+		to_chat(user, SPAN_WARNING("You need to grip someone first!"))
+		return
+	. = ..()
+	if(!.)
+		return
+	user.visible_message(SPAN_WARNING("[user] squeezes [user.get_pronoun("his")] hand!"), SPAN_WARNING("You squeeze your hand to tighten the psionic force around [victim]."))
+	to_chat(victim, SPAN_DANGER("<font size=4>You are crushed by an invisible force!</font>"))
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/grip.dm:33]:
```suggestion
	to_chat(victim, SPAN_DANGER(FONT_HUGE("You are crushed by an invisible force!")))
```

File: code/modules/psionics/abilities/grip.dm [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	if(victim)
+		to_chat(user, SPAN_WARNING("You can't target more than one person with this!"))
+		return
+	. = ..()
+	if(!.)
+		return
+	var/mob/living/M = hit_atom
+	user.visible_message(SPAN_DANGER("[user] extends [user.get_pronoun("his")] arm and makes a grab motion towards [M]!"),
+						SPAN_DANGER("You extend your arm and grab [M] with your psionic energy!"))
+	to_chat(M, SPAN_DANGER("<font size=4>You feel an invisible force tighten around you!</font>"))
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/grip.dm:48]:
```suggestion
	to_chat(M, SPAN_DANGER(FONT_HUGE("You feel an invisible force tighten around you!")))
```

File: code/modules/psionics/abilities/mend.dm [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		return
+
+	user.visible_message(SPAN_NOTICE("[user] lays a palm on [H]..."), SPAN_NOTICE("You lay your palm on [H] and get to work."))
+	to_chat(user, SPAN_NOTICE("You start by flowing your regenerative psionic energy through their vital organs..."))
+	for(var/obj/item/organ/O in H.internal_organs)
+		if(do_mob(user, H, 15 SECONDS))
+			if(O.damage > 0) // Fix internal damage
+				to_chat(SPAN_NOTICE("You mend their [O]'s bruising."))
+				O.heal_damage(O.damage)
+			if(O.status & ORGAN_BROKEN)
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/mend.dm:42]:
```suggestion
			if(HAS_FLAG(O.status, ORGAN_BROKEN))
```

File: code/modules/psionics/abilities/mend.dm [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+			O.heal_damage(0, O.damage, internal = TRUE, robo_repair = FALSE)
+
+	to_chat(user, SPAN_NOTICE("Finally, you flowing your regenerative psionic energy through their broken limbs..."))
+	for(var/obj/item/organ/E in H.bad_external_organs) // Fix bones
+		var/obj/item/organ/external/affected = E
+		if(do_mob(user, H, 10 SECONDS))
+			if((affected.damage < affected.min_broken_damage * config.organ_health_multiplier) && (affected.status & ORGAN_BROKEN))
+				to_chat(SPAN_NOTICE("You mend their [affected] together."))
+				affected.status &= ~ORGAN_BROKEN
+			if(affected.status & ORGAN_ARTERY_CUT)
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/mend.dm:63]:
```suggestion
			if(HAS_FLAG(affected.status, ORGAN_ARTERY_CUT))
```

File: code/modules/psionics/abilities/mend.dm [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+				to_chat(SPAN_NOTICE("You mend a spliced artery in their [affected]."))
+				affected.status &= ~ORGAN_ARTERY_CUT
+			if(affected.status & ORGAN_DEAD)
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/mend.dm:66]:
```suggestion
			if(HAS_FLAG(affected.status, ORGAN_DEAD))
```

File: code/modules/psionics/abilities/psi_search.dm [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+					continue
+				level_humans |= H
+				if(H.psi)
+					if(H.psi.get_rank() >= PSI_RANK_APEX)
+						found_apex = TRUE
+		if(!length(level_humans))
+			to_chat(L, SPAN_WARNING("The Nlom is quiet and empty here."))
+			return TRUE
+		if(found_apex)
+			to_chat(L, SPAN_DANGER("<font size=4>You reach out into the Nlom and your senses are overwhelmed by a massive signature!</font>"))
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/psi_search.dm:44]:
```suggestion
			to_chat(L, SPAN_DANGER(FONT_HUGE("You reach out into the Nlom and your senses are overwhelmed by a massive signature!")))
```

File: code/modules/psionics/abilities/shockwave.dm [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	. = ..()
+	if(!.)
+		return
+
+	var/user_dir = user.dir
+	var/turf/T = get_turf(user)
+	if(isspaceturf(T))
+		to_chat(user, SPAN_WARNING("You charge your shockwave, slam your foot down... and then remember that you're in space."))
+		return
+	user.visible_message(SPAN_DANGER("<font size=4>[user] charges [user.get_pronoun("his")] foot with psionic energy and slams it down!</font>"),
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/shockwave.dm:28]:
```suggestion
	user.visible_message(SPAN_DANGER(FONT_HUGE("[user] charges [user.get_pronoun("his")] foot with psionic energy and slams it down!")),
```

File: code/modules/psionics/abilities/shockwave.dm [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+						SPAN_DANGER("<font size=4>You charge your foot with psionic energy and slam it down!</font>"))
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/shockwave.dm:29]:
```suggestion
						SPAN_DANGER(FONT_HUGE("You charge your foot with psionic energy and slam it down!")))
```

File: code/modules/psionics/abilities/singularity.dm [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+		return
+	. = ..()
+	if(!.)
+		return
+
+	var/obj/effect/singularity/S = new(get_turf(hit_atom))
+	S.creator = user
+	singularity = S
+	playsound(get_turf(user), 'sound/effects/fingersnap.ogg')
+	user.visible_message(SPAN_DANGER("<font size=4>[user] snaps [user.get_pronoun("his")] fingers and generates a hole of psionic energy!</font>"),
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/singularity.dm:34]:
```suggestion
	user.visible_message(SPAN_DANGER(FONT_HUGE("[user] snaps [user.get_pronoun("his")] fingers and generates a hole of psionic energy!")),
```

File: code/modules/psionics/abilities/time_stop.dm [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+		L.silent = 0
+	return ..()
+
+/obj/item/spell/time_stop/on_use_cast(mob/user)
+	. = ..()
+	if(!.)
+		return
+
+	if(do_after(user, 1 SECOND))
+		user.visible_message(SPAN_DANGER("<font size=4>[user] extends [user.get_pronoun("his")] arms to [user.get_pronoun("his")] sides!</font>"),
```
Comment (FluffyGhoster) [inline code/modules/psionics/abilities/time_stop.dm:33]:
```suggestion
		user.visible_message(SPAN_DANGER(FONT_HUGE("[user] extends [user.get_pronoun("his")] arms to [user.get_pronoun("his")] sides!")),
```

File: code/modules/psionics/events/mini_spasm.dm [hunk 18]
```diff
@@ -48,16 +48,12 @@
...
 		to_chat(victim, SPAN_DANGER("An indescribable, brain-tearing sound hisses from [icon2html(source, victim)] \the [source], and you collapse in a seizure!"))
 		victim.seizure()
-		var/new_latencies = rand(2,4)
-		var/list/faculties = list(PSI_COERCION, PSI_REDACTION, PSI_ENERGISTICS, PSI_PSYCHOKINESIS)
-		for(var/i = 1 to new_latencies)
-			to_chat(victim, SPAN_DANGER("<font size = 3>[pick(psi_operancy_messages)]</font>"))
-			victim.adjustBrainLoss(rand(10,20))
-			victim.set_psi_rank(pick_n_take(faculties), 1)
-			sleep(30)
+		to_chat(victim, SPAN_DANGER("<font size = 3>[pick(psi_operancy_messages)]</font>"))
```
Comment (FluffyGhoster) [inline code/modules/psionics/events/mini_spasm.dm:51]:
```suggestion
		to_chat(victim, SPAN_DANGER(FONT_LARGE("[pick(psi_operancy_messages)]")))
```

File: code/modules/psionics/interface/ui_hub.dm [hunk 19]
```diff
@@ -30,25 +29,68 @@
...
+
+/obj/screen/psi/hub/ui_data(mob/user)
+	var/list/data = list()
+	var/owner_rank = owner.psi.get_rank()
+	data["available_psionics"] = list()
+	data["psi_rank"] = psychic_ranks_to_strings[owner_rank]
+	data["psi_points"] = owner.psi.psi_points
+	data["bought_powers"] = owner.psi.psionic_powers
+	for(var/singleton/psionic_power/P in GET_SINGLETON_SUBTYPE_LIST(/singleton/psionic_power))
+		if((P.ability_flags & PSI_FLAG_SPECIAL) && !(P.type in owner.psi.psionic_powers))
```
Comment (FluffyGhoster) [inline code/modules/psionics/interface/ui_hub.dm:59]:
```suggestion
		if(HAS_FLAG(P.ability_flags, PSI_FLAG_SPECIAL) && !(P.type in owner.psi.psionic_powers))
```

File: code/modules/psionics/interface/ui_hub.dm [hunk 20]
```diff
@@ -30,25 +29,68 @@
...
+			continue
+		if(owner_rank < P.minimum_rank)
+			continue
+		/// Apex and Limitless abilities are automatically given, but we want them to have said abilities in the point shop so the users know what they do.
+		if(owner_rank < PSI_RANK_APEX && (P.ability_flags & PSI_FLAG_APEX))
```
Comment (FluffyGhoster) [inline code/modules/psionics/interface/ui_hub.dm:64]:
```suggestion
		if(owner_rank < PSI_RANK_APEX && HAS_FLAG(P.ability_flags, PSI_FLAG_APEX))
```

File: code/modules/psionics/interface/ui_hub.dm [hunk 21]
```diff
@@ -30,25 +29,68 @@
...
+			continue
+		if(owner_rank < PSI_RANK_LIMITLESS && (P.ability_flags & PSI_FLAG_LIMITLESS))
```
Comment (FluffyGhoster) [inline code/modules/psionics/interface/ui_hub.dm:66]:
```suggestion
		if(owner_rank < PSI_RANK_LIMITLESS && HAS_FLAG(P.ability_flags, PSI_FLAG_LIMITLESS))
```

File: code/modules/psionics/interface/ui_hub.dm [hunk 22]
```diff
@@ -30,25 +29,68 @@
...
+			continue
+		if(!(P.ability_flags & PSI_FLAG_CANON))
```
Comment (FluffyGhoster) [inline code/modules/psionics/interface/ui_hub.dm:68]:
```suggestion
		if(NOT_FLAG(P.ability_flags, PSI_FLAG_CANON))
```

File: code/modules/psionics/interface/ui_hub.dm [hunk 23]
```diff
@@ -30,25 +29,68 @@
...
+			if(owner_rank < PSI_RANK_HARMONIOUS && (P.ability_flags & PSI_FLAG_EVENT))
```
Comment (FluffyGhoster) [inline code/modules/psionics/interface/ui_hub.dm:69]:
```suggestion
			if(owner_rank < PSI_RANK_HARMONIOUS && HAS_FLAG(P.ability_flags, PSI_FLAG_EVENT))
```

File: code/modules/psionics/interface/ui_hub.dm [hunk 24]
```diff
@@ -30,25 +29,68 @@
...
+				continue
+		if(owner_rank < PSI_RANK_HARMONIOUS && (P.ability_flags & PSI_FLAG_ANTAG))
```
Comment (FluffyGhoster) [inline code/modules/psionics/interface/ui_hub.dm:71]:
```suggestion
		if(owner_rank < PSI_RANK_HARMONIOUS && HAS_FLAG(P.ability_flags, PSI_FLAG_ANTAG))
```

**Review by FluffyGhoster — COMMENTED** [review]
some macro usages

File: code/modules/client/preferences.dm [hunk 25]
```diff
@@ -471,6 +473,12 @@ var/list/preferences_datums = list()
 
 	character.headset_choice = headset_choice
 
+	if(length(psionics) && character.client)
+		for(var/power in psionics)
+			var/singleton/psionic_power/P = GET_SINGLETON(text2path(power))
+			if(istype(P) && (P.ability_flags & PSI_FLAG_CANON))
```
[the remark this answers is not in the document]
Comment (NonQueueingMatt) [reply code/modules/client/preferences.dm:479]: this is literally just an operation, it doesn't need to be a macro

**Review by SleepyGemmy — APPROVED** [review]

**Pull request merged**